r1cs/src/test_constraint_checker.rs

```diff
@@ -0,0 +1,173 @@
+    }
+
+    #[inline]
+    pub fn is_satisfied(&self) -> bool {
+        !self.found_unsatisfactory_constraint
+    }
+
+    #[inline]
+    pub fn num_constraints(&self) -> usize {
+        self.num_constraints
+    }
+}
+
+impl<F: Field> ConstraintSystem<F> for TestConstraintChecker<F> {
+    type Root = Self;
+
+    fn alloc<Fn, A, AR>(&mut self, _annotation: A, f: Fn) -> Result<Variable, SynthesisError>
+    where
+        Fn: FnOnce() -> Result<F, SynthesisError>,
+        A: FnOnce() -> AR,
+        AR: AsRef<str>,
+    {
+        let index = self.private_variables.len();
+        self.private_variables.push(f()?);
+        let var = Variable::new_unchecked(Index::Private(index));
+
+        Ok(var)
+    }
+
+    fn alloc_input<Fn, A, AR>(&mut self, _annotation: A, f: Fn) -> Result<Variable, SynthesisError>
+    where
+        Fn: FnOnce() -> Result<F, SynthesisError>,
+        A: FnOnce() -> AR,
+        AR: AsRef<str>,
+    {
+        let index = self.public_variables.len();
+        self.public_variables.push(f()?);
+        let var = Variable::new_unchecked(Index::Public(index));
+
+        Ok(var)
+    }
+
+    fn enforce<A, AR, LA, LB, LC>(&mut self, annotation: A, a: LA, b: LB, c: LC)
+    where
+        A: FnOnce() -> AR,
+        AR: AsRef<str>,
+        LA: FnOnce(LinearCombination<F>) -> LinearCombination<F>,
+        LB: FnOnce(LinearCombination<F>) -> LinearCombination<F>,
+        LC: FnOnce(LinearCombination<F>) -> LinearCombination<F>,
+    {
+        self.num_constraints += 1;
+
+        let eval_lc = |lc: Vec<(Variable, F)>| -> F {
+            lc.into_iter()
+                .map(|(var, coeff)| {
+                    let value = match var.get_unchecked() {
+                        Index::Public(index) => self.public_variables[index],
+                        Index::Private(index) => self.private_variables[index],
+                    };
+                    value * coeff
+                })
+                .sum::<F>()
+        };
+
+        let a = eval_lc(a(LinearCombination::zero()).0);
+        let b = eval_lc(b(LinearCombination::zero()).0);
+        let c = eval_lc(c(LinearCombination::zero()).0);
+
+        if a * b != c {
+            self.found_unsatisfactory_constraint = true;
+
+            if self.first_unsatisfied_constraint.is_none() {
+                let new = annotation().as_ref().to_string();
+                assert!(!new.contains('/'), "'/' is not allowed in names");
+
+                let mut path = self.segments.clone();
+                path.push(new);
+                self.first_unsatisfied_constraint = Some(path.join("/"));
+            }
+        }
+    }
+
+    fn push_namespace<NR: AsRef<str>, N: FnOnce() -> NR>(&mut self, name_fn: N) {
+        let new = name_fn().as_ref().to_string();
+        assert!(!new.contains('/'), "'/' is not allowed in names");
+
+        self.segments.push(new)
+    }
+
+    fn pop_namespace(&mut self) {
+        self.segments.pop();
```

Comment:
doesn't the memory grow a lot if the variables associated with a namespace aren't removed from their respective vectors when a namespace is popped?

Reply:
if each segment was associated with 2 values corresponding to the number of private and public variables, they could be used to truncate their respective vectors; this is under the assumption that once a namespace segment is popped, the variables that were declared within it can go out of scope.

Reply:
This might be less an issue here, as (1) the segments are stored only in one place, and (2) the segment is not effectively associated with the counts of private and public variables, as there is only a global count of them.

Reply:
(which also means that this checker now offers much less functionality for looking into the cost, so it may be just a checker for the `test`)